### PR TITLE
Implement the `double_circle` shape including basic shape and the sketch.

### PR DESCRIPTION
**Pull request by OneRain233**
Implement the `double_circle` shape including basic shape and the sketch.
#564

### CODE REVIEW

File: lib/shape/shape.go
Comment (pleshevskiy) [inline lib/shape/shape.go:173]: You broke the ascii diagram :)

**Comment by ejulio-ts** [comment]
@OneRain233 can you include an image of your change? Quite hard to guess what it is from changes :)

Also, can you update this test case https://github.com/terrastruct/d2/blob/master/e2etests/stable_test.go#L26?

**Comment by OneRain233** [comment]
For the `all_shape` test, I found that the size of the double circle shape is larger than others. But when I ran the normal `.2d` file, it is normal. I dont know the reason.
```
x1:{shape:double_circle}
x2:{shape:circle}
x1 -> x2
```
Screenshot from 2022-12-30 21-35-31


test

**Comment by pleshevskiy** [comment]
@OneRain233 The shape stretches from the label size. I think you can make a little less internal padding

**Comment by OneRain233** [comment]
> @OneRain233 The shape stretches from the label size. ~I think you can make a little less internal padding~

Maybe I know the reasons, because the name `double_circle` is longer than others name.

File: d2renderers/d2sketch/sketch.go
Comment (ejulio-ts) [inline d2renderers/d2sketch/sketch.go:137]: go styling is `camelCase` and not `snake_case`

File: d2renderers/d2svg/d2svg.go
Comment (ejulio-ts) [inline d2renderers/d2svg/d2svg.go:501]: you can call `renderOval` twice here.
another alternative is to remove this function and call `renderOval` twice in the main one

**Review by ejulio-ts — COMMENTED** [review]
cool!
Besides updating the test cases, you need to update their output too
```
TESTDATA_ACCEPT=1 ./ci/test.sh
```
and it'll update the SVGs and JSONs of the test cases, otherwise it'll break during the CI run


I'll let @alixander or @nhooyr make the final call on this one.

**Review by alixander — CHANGES_REQUESTED** [review]
hi @OneRain233 thanks for taking the initiative here! 

I don't think a separate shape is right for it though. There'd no 3d_square or multiple_circle or blue_oval; these are just attributes in D2. If this were to be implemented, it should be a style like `double: true` or `double-border: true`.

I don't personally know what this double-border style is for, but I've seen it used, and, recently in this: http://www.bpmb.de/images/BPMN2_0_Poster_EN.pdf . So I do think it's good to offer this in D2. 

Btw, is there anything you felt was missing from the CONTRIBUTING.md? (did you use it at all? 😃 )

**Comment by OneRain233** [comment]
> hi @OneRain233 thanks for taking the initiative here!
> 
> I don't think a separate shape is right for it though. There'd no 3d_square or multiple_circle or blue_oval; these are just attributes in D2. If this were to be implemented, it should be a style like `double: true` or `double-border: true`.
> 
> I don't personally know what this double-border style is for, but I've seen it used, and, recently in this: http://www.bpmb.de/images/BPMN2_0_Poster_EN.pdf . So I do think it's good to offer this in D2. Oh I guess I've seen it in state machines too
> 
> Btw, is there anything you felt was missing from the CONTRIBUTING.md? (did you use it at all? smiley )

Yes, I just want to use the double circle in the accept state of state machine.

**Comment by alixander** [comment]
@OneRain233 how about `double-border: true | false`? not `double`, to avoid confusion with `multiple`

**Comment by OneRain233** [comment]
> @OneRain233 how about `double-border: true | false`? not `double`, to avoid confusion with `multiple`
Yes, `double-border: true | false` is better.

**Comment by OneRain233** [comment]
Should the tag `double-border` be used only for circles or for all shapes? It may be strange for some shapes to use this tag?  @alixander

**Comment by OneRain233** [comment]
I have made `double-border` for the `circle` shape only.

**Comment by alixander** [comment]
@OneRain233 I think for circles, ovals, squares, and rectangles.

**Comment by OneRain233** [comment]
Screenshot from 2022-12-31 15-58-41
Now, it looks like this.

File: d2renderers/d2svg/d2svg.go
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:497]: this should be a constant like https://github.com/terrastruct/d2/blob/bd15db4fdf77a6a1ef469489ffbee89089b83a17/d2renderers/d2svg/d2svg.go#L46

File: d2renderers/d2svg/d2svg.go
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:664]: i'd prefer the true statement to come first in if/else

File: d2renderers/d2svg/d2svg.go
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:680]: we should remove this. i don't think it'd ever look good for `multiple` + `double-border`. we just choose one at random and say it takes precedence.

**Review by alixander — CHANGES_REQUESTED** [review]
sorry for the delay in reviewing this. it looks like all the tests changed because you ran on go 1.19. Can you run with go 1.18 please?

File: d2renderers/d2svg/d2svg.go
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:47]: should move this to `d2target`, so it can be used both here and `d2sketch` without import cycle issues

File: d2renderers/d2sketch/sketch.go
Comment (alixander) [inline d2renderers/d2sketch/sketch.go:127]: use constant defined

File: d2renderers/d2sketch/sketch.go
Comment (alixander) [inline d2renderers/d2sketch/sketch.go:142]: use constant defined

File: d2renderers/d2sketch/sketch.go
Comment (alixander) [inline d2renderers/d2sketch/sketch.go:189]: why's this number different?

File: d2renderers/d2svg/d2svg.go
Comment (alixander) [inline d2renderers/d2svg/d2svg.go:723]: why 15?

**Review by alixander — CHANGES_REQUESTED** [review]
this turned out to look really nice, especially in sketch mode! Just a few more things:

Looks like you also need to account for when pad is 0.
`d2 --pad 0 input.d2` gave me this:
Screen Shot 2023-01-21 at 2 32 26 PM

You can see here for example of fixing: https://github.com/terrastruct/d2/pull/685

Also just more merge conflicts with tests and changelog, sorry about that.

**Comment by OneRain233** [comment]
> this turned out to look really nice, especially in sketch mode! Just a few more things:
> 
> Looks like you also need to account for when pad is 0. `d2 --pad 0 input.d2` gave me this: Screen Shot 2023-01-21 at 2 32 26 PM
> 
> You can see here for example of fixing: #685
> 
> Also just more merge conflicts with tests and changelog, sorry about that.

Should I sync from the original repo and then fix the issue?

**Comment by ejulio-ts** [comment]
@OneRain233, yes.
You merge/rebase the current `master` and update the tests.

**Comment by alixander** [comment]
@OneRain233 looks like all the commits were pulled in

**Comment by OneRain233** [comment]
I think the issue #685 is solved. Now it looks like this.
test
test1

**Review by alixander — CHANGES_REQUESTED** [review]
ready to merge after 1 minor thing.

**Comment by alixander** [comment]
looks like one more merge + test regeneration plz @OneRain233

**Review by alixander — APPROVED** [review]
we'll just merge and fix the conflict in another commit on master. Thank you @OneRain233 ! Sorry for the delays in this one

**Comment by nhooyr** [comment]
@alixander note for the future, github normally automatically gives you write access to a contributor's branch unless the contributor has declined your access.

**Pull request merged**
